### ISSUE TITLE
FR: Firefox port?

### ISSUE THREAD

**Issue by Offerel** [qownnotes/web-companion#1]
Could you please also add a Firefox port? Mozillas Webextension API is much similar to Chrome.

**Comment by pbek** [comment]
Would be nice, but currently I don't plan to. The Chrome one was already quite some work...

**Comment by Offerel** [comment]
I know. Developing Webextensions can frustrating. If you don't have a problem with it, i could try to fork this repo tomorrow and try to migrate it to Firefox. I don't know if i can get it to work, since i have only developed 1 extension yet. But i can try it.

**Comment by pbek** [comment]
Sure, let me know what you come up with! 👍

**Comment by pbek** [comment]
Great work! I have commented on https://github.com/qownnotes/chrome-web-companion/pull/2.

> <all_urls>

on Chrome there was a new permission for the current tab that went around this

> I don't know, if you want to release it to Mozillas AddOn Page or i should do it.

I'll first test the extension and take a look at the shop processes for Firefox.

> Would you like to migrate my syncing AddOn to Chrome?

I use Nextcloud bookmarks :)

**Comment by Offerel** [comment]
> > <all_urls>
> on Chrome there was a new permission for the current tab that went around this

Yes i know, but if i use that permission, the function isn't working. According to the developer network <all_urls> is needed to get it to work. But i can try to find a other way.

> I use Nextcloud bookmarks :)
I hate Nextcloud :-) It's so damn slow on a Raspberry Pi 3B+. There are so many unneeded SQL query's which slow it down. In theory Nextcloud is a wonderful thing, but in reality it's neither fish nor meat. The advantage of my AddOn is it, that you can use it with every PHP enabled webserver server or as alternative, every WebDAV Share. It's not bound to Nextcloud (but you can use it also with Nextcloud). And yes, you can run Firefox Sync in parallel.

**Comment by Offerel** [comment]
> > <all_urls>
> on Chrome there was a new permission for the current tab that went around this
OK, i tested a little bit more. It seems that in Firefox it is only possible with the <all_urls> permission according to https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/captureVisibleTab

But adding this permission does not have any negative side effect for Chrome users.

I have tested the AddOn now in both browsers, Chrome and Firefox. It works in both. Only in Chrome i get error/warning, since i have added the "applications" key in manifest.json. This key is mandatory in Firefox for some functions, but not recognized in Chrome. However, it doesn't make problems in Chrome. It seems that adding the popup, also fixed a small bug in your version, which doesn't display the QOwnNotes.png in the menu (it was grayed out after installing the extension from Chrome Store).

If it is no problem for you to maintain a separate manifest.json for every browser, it seems to be ok to use 1 repository for both extensions. 

So the only real difference is the manifest,json.

**Comment by pbek** [comment]
I guess I will implement a script that creates ZIP files for Chrome (zip) and Firefox (xpi).

**Comment by cadeyrn** [comment]
Maybe https://github.com/webextension-toolbox/webextension-toolbox/blob/master/README.md can help you.

**Comment by pbek** [comment]
> But adding this permission does not have any negative side effect for Chrome users.

none beside making people nervous ;)

> Maybe https://github.com/webextension-toolbox/webextension-toolbox/blob/master/README.md helps.

thank you, @cadeyrn, I heard of it before and will look more into it.

**Comment by Offerel** [comment]
@pbek yes, i know. That's why I explain for my extension every permission in detail and why the extension needs exactly this permission. I think that's all we can do.

@cadeyrn good point. I remember you have mentioned this before to me in your forum :smile: and what should I say, I use it now sometimes. I hated it before. But it's useful indeed.

**Comment by pbek** [comment]
> That's why I explain for my extension every permission in detail and why the extension needs exactly this permission. 

sure, I was only talking about Chrome, where the permission isn't needed

**Comment by pbek** [comment]
@cadeyrn, haha, getting everything to work again with webpack (used by webextension-toolbox) is really a hell :laughing: 

Currently I'm at https://github.com/qownnotes/web-companion/blob/6a40f2cf58bbfc6753c90f226cb705176fdb5684/app/scripts/background.js#L145-L154 to import the `functions.js` via webpack to make fetching the contet of a whole page working again...

**Comment by pbek** [comment]
Yay, the extension is now in the Firefox store: <https://addons.mozilla.org/en-US/firefox/addon/qownnotes-web-companion/> :tada: Thank you all for your help!

**Comment by Offerel** [comment]
Great! Many thx for releasing.

**Comment by pbek** [comment]
It took a looooot of effort, but I learned quite some stuff in the process. :smile: 

Does the extension from the store work for you?

**Comment by Offerel** [comment]
Yes it does. I have send you additionally a larger e-mail in german. I will test this addon tomorrow in more detail, but for now it works. I can create notes, as whole page and snippet. If i found something, i can make a new issue and if i have the time, make a pull request with a fix.

**Comment by pbek** [comment]
Thank you, I answered ;)

**Comment by pbek** [comment]
btw. there also is a Telegram group, if you want to chat there... https://t.me/QOwnNotes

**Comment by pbek** [comment]
Now I know what happens if you add `<all_urls>` as `permission` (for the new "bookmark all tabs" feature) and post it in the chrome store... It triggers a "several workdays approval" process instead of instant approval. :thinking:

**Comment by Offerel** [comment]
Google :-)

**Comment by Offerel** [comment]
Also in Firefox? Great, i will look in to this...

**Comment by pbek** [comment]
I didn't test if `tabs` will also work for screenhots, but you need it in both browsers for getting the urls of all open tabs.

**Comment by pbek** [comment]
But accessing bookmarks with the extension works like a charm with the extension. I love it a lot. :smile:
